File: creational/singleton/typescript/src/realLife/EventService.ts
type Listener = (value: string) => void;

type Unsubscribe = { unsubscribe: () => void };

class EventService {
  private listeners: Listener[] = [];

  protected static _instance: EventService | null = null;

  private constructor() {}

  static get instance(): EventService {
    if (EventService._instance === null) {
      EventService._instance = new EventService();
    }

    return EventService._instance;
  }

  public emit(value: string) {
    this.notifyListeners(value);
  }

  public subscribe(listener: Listener): Unsubscribe {
    this.listeners.push(listener);
    return {
      unsubscribe: () => {
        this.listeners = this.listeners.filter(
          (anotherListener) => anotherListener !== listener
        );
      },
    };
  }

  private notifyListeners(value: string) {
    this.listeners.forEach((listener) => listener(value));
  }
}

const eventService = EventService.instance;

const eventSubscription = eventService.subscribe((event) => console.log(event));

eventService.emit("First Event");

EventService.instance.emit("Second Event");

eventSubscription.unsubscribe();

EventService.instance.emit("Third event"); // Will not emit

File: design-patterns/creational/singleton/dart/realLife/EventService.dart
typedef Listener = void Function(String value);

class Unsubscribe {
  final void Function() unsubscribe;
  Unsubscribe({required this.unsubscribe});
}

class EventService {
  static final EventService _singleton = EventService._internal();

  factory EventService() {
    return _singleton;
  }

  EventService._internal();

  final List<Listener> _listeners = [];

  void emit(String value) {
    notifyListeners(value);
  }

  Unsubscribe subscribe(Listener listener) {
    _listeners.add(listener);
    return Unsubscribe(unsubscribe: () => _listeners.remove(listener));
  }

  void notifyListeners(String value) {
    _listeners.forEach((listener) => listener(value));
  }
}
